diff -up rdesktop-1.6.0/scard.c.jx rdesktop-1.6.0/scard.c
--- rdesktop-1.6.0/scard.c.jx	2008-02-14 06:37:18.000000000 -0500
+++ rdesktop-1.6.0/scard.c	2012-08-22 16:51:43.285300485 -0400
@@ -1323,7 +1323,7 @@ TS_SCardLocateCardsByATR(STREAM in, STRE
 				if (equal)
 				{
 					rsCur->dwEventState |= 0x00000040;	/* SCARD_STATE_ATRMATCH 0x00000040 */
-					memcpy(ResArray + j, rsCur, sizeof(SCARD_READERSTATE_A));
+					memcpy(ResArray + j, rsCur, sizeof(SERVER_SCARD_READERSTATE_A));
 					DEBUG_SCARD(("SCARD:    \"%s\"\n",
 						     rsCur->szReader ? rsCur->szReader : "NULL"));
 					DEBUG_SCARD(("SCARD:        user: 0x%08x, state: 0x%08x, event: 0x%08x\n", (unsigned) rsCur->pvUserData, (unsigned) rsCur->dwCurrentState, (unsigned) rsCur->dwEventState));
