diff -up rdesktop-1.7.1/scard.c~ rdesktop-1.7.1/scard.c
--- rdesktop-1.7.1/scard.c~	2011-12-07 04:49:44.000000000 -0500
+++ rdesktop-1.7.1/scard.c	2013-06-24 23:27:01.000000000 -0400
@@ -1354,7 +1354,7 @@ TS_SCardLocateCardsByATR(STREAM in, STRE
 				if (equal)
 				{
 					rsCur->dwEventState |= 0x00000040;	/* SCARD_STATE_ATRMATCH 0x00000040 */
-					memcpy(ResArray + j, rsCur, sizeof(SCARD_READERSTATE));
+					memcpy(ResArray + j, rsCur, sizeof(SERVER_SCARD_READERSTATE_A));
 					DEBUG_SCARD(("SCARD:    \"%s\"\n",
 						     rsCur->szReader ? rsCur->szReader : "NULL"));
 					DEBUG_SCARD(("SCARD:        user: %p, state: 0x%08x, event: 0x%08x\n", rsCur->pvUserData, (unsigned) rsCur->dwCurrentState, (unsigned) rsCur->dwEventState));
